src/client.rs

```diff
@@ -0,0 +1,164 @@
+use std::borrow::Cow;
+
+use tokio::sync::mpsc::{self, channel, Sender};
+use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::ReceiverStream;
+use tokio_stream::Stream;
+
+use crate::background::{BackgroundTask, Command};
+use crate::errors::Error;
+
+/// Basic synchronization client enabling one to send signals, await barriers and subscribe or publish to a topic.
+pub struct Client {
+    cmd_tx: Sender<Command>,
+    handle: JoinHandle<()>,
+}
+
+impl Drop for Client {
+    fn drop(&mut self) {
+        self.handle.abort();
+    }
+}
+
+impl Client {
+    pub async fn new() -> Result<Self, Box<dyn std::error::Error>> {
+        let (cmd_tx, cmd_rx) = channel(10);
+
+        let mut background = BackgroundTask::new(cmd_rx).await?;
+
+        let handle = tokio::spawn(async move {
+            background.run().await;
+        });
+
+        Ok(Self { cmd_tx, handle })
+    }
+
+    pub async fn publish(
+        &self,
+        topic: impl Into<Cow<'static, str>>,
+        payload: impl Into<Cow<'static, str>>,
+    ) -> Result<u64, Error> {
+        let (sender, receiver) = oneshot::channel();
+
+        let cmd = Command::Publish {
+            topic: topic.into().into_owned(),
+            payload: payload.into().into_owned(),
+            sender,
+        };
+
+        self.cmd_tx.send(cmd).await.expect("receiver not dropped");
+
+        receiver.await.expect("sender not dropped")
+    }
+
+    pub async fn subscribe(
+        &self,
+        topic: impl Into<Cow<'static, str>>,
+    ) -> impl Stream<Item = Result<String, Error>> {
+        let (stream, out) = mpsc::channel(10);
+
+        let cmd = Command::Subscribe {
+            topic: topic.into().into_owned(),
+            stream,
+        };
+
+        self.cmd_tx.send(cmd).await.expect("receiver not dropped");
+
+        ReceiverStream::new(out)
+    }
+
+    pub async fn signal_and_wait(
+        &self,
+        state: impl Into<Cow<'static, str>>,
+        target: u64,
+    ) -> Result<u64, Error> {
+        let (sender, receiver) = oneshot::channel();
+
+        let state = state.into().into_owned();
+
+        let cmd = Command::Signal {
+            state: state.clone(),
+            sender,
+        };
+
+        self.cmd_tx.send(cmd).await.expect("receiver not dropped");
```

Comment:
Could this be de-duplicated by calling `self.signal(...)` instead?

Reply:
will do!